### PR TITLE
add color-field 2D renderer

### PR DESCRIPTION
**Pull request by bdeket**
Among other things allows to make plots of fractals.
It is similar to contour-intervals, but for erratic functions (like fractals) I don't know how to get contour-intervals to display what I want.

----

Sample code:

```racket
#lang typed/racket
(require plot)
;; https://rosettacode.org/wiki/Mandelbrot_set#Racket
(define (iterations [a : Number][z : Number][i : Nonnegative-Integer]) : Nonnegative-Integer
  (define z′ (+ (* z z) a))
  (if (or (<= 255 i) (< 2 (magnitude z′)))
      i
      (iterations a z′ (+ i 1))))

(define (iter->color [i : Nonnegative-Integer]) : Plot-Color
  (if (= i 255)
      'black
      (list (* 5 (modulo i 15)) (* 32 (modulo i 7)) (* 8 (modulo i 31)))))

(define (mandelbrot [x : Real][y : Real]) : Plot-Color
  (define z (make-rectangular (* 1.0 x) (* 1.0 y)))
  (iter->color (iterations z 0 0)))

(plot (color-field mandelbrot -2.25 0.75 -1.25 1.25 #:samples 200))
```

Produces:

mandelbrot-plot

### CODE REVIEW

**Comment by alex-hhh** [comment]
> Among other things allows to make plots of fractals.

Allow me to ask the reverse question.  Except for plots of fractals, what other things can `color-field` be used for?  Also, even for fractals, is there a particular need for displaying fractals on a plot surface?  You could generate fractals using pictures (e.g using the `dc` pict constructor).

File: plot-lib/plot/private/plot2d/color-field.rkt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        (list (* 5 (modulo i 15)) (* 32 (modulo i 7)) (* 8 (modulo i 31)))))
+
+  (define (mandelbrot [x : Real][y : Real]) : Plot-Color
+    (define z (make-rectangular (* 1.0 x) (* 1.0 y)))
+    (iter->color (iterations z 0 0)))
+
+  (plot (color-field mandelbrot -2.25 0.75 -1.25 1.25
+                     #:samples 200)))
+
+#;(module+ test
```
Comment (alex-hhh) [inline plot-lib/plot/private/plot2d/color-field.rkt:98]: Can you please add this as a test to the `plot-test` package (there are some examples on how to do that, but the tests essentially check that the plot images are generated correctly).  Let me know if you have trouble with this.

**Comment by bdeket** [comment]
Regarding fractals:
the nice thing of doing it with plot is that zooming is there by default.

Other things that would be possible:
instead of making a vector-field, use color intensity for the length of the vector, and color for the angle
(I would not know how to do that with contours)

What I am mainly using it for:
making error plots of complex functions. Where the errors can be really erratic. and the zooming function of plot gives me the option to explore where the problems are

File: plot-lib/plot/private/common/parameters.rkt [hunk 2]
```diff
@@ -228,6 +228,12 @@
 (defparam candlestick-line-style Plot-Pen-Style 'solid)
 (defparam2 candlestick-alpha Real Nonnegative-Real 2/3 (unit-ivl 'candlestick-alpha))
 
+
+;; color fields
+
+(defparam color-field-samples Positive-Integer 20)
+(defparam2 color-field-alpha Real Nonnegative-Real 1 (unit-ivl 'color-field-alpha))
```
[the remark this answers is not in the document]
Comment (bdeket) [reply plot-lib/plot/private/common/parameters.rkt:235]: Could you tell me where would I do this? I don't find any corresponding group for vector-field-samples or vector-field-alpha...

[the remark this answers is not in the document]
Comment (alex-hhh) [reply plot-lib/plot/private/common/parameters.rkt:235]: Actually, looking at the implementation, you don't need to do this, as you use the parameters as the default values for the `color-field` parameters.

File: plot-doc/plot/scribblings/params.scrbl [hunk 3]
```diff
@@ -310,6 +310,13 @@ width should take that into consideration. For example, a width of 86400 may be
 as there are 86400 seconds in a day. This candle will be exactly one day in width.
 }
 
+@section{Color fields}
+
+@deftogether[((defparam color-field-samples n exact-positive-integer? #:value 20)
+              (defparam error-bar-alpha alpha (real-in 0 1) #:value 1))]{
+The default sample rate and opacity used by @racket[color-field].
```
Comment (alex-hhh) [inline plot-doc/plot/scribblings/params.scrbl:317]: Can you please add a history entry to these variables:

```
@history[#:added "7.9"]
```

File: plot-doc/plot/scribblings/renderer2d.scrbl [hunk 4]
```diff
@@ -198,6 +198,21 @@ fourth, and fifth elements in each vector comprise the open, high, low, and clos
...
+          [y-min (or/c rational? #f) #f] [y-max (or/c rational? #f) #f]
+          [#:samples samples exact-positive-integer? (color-field-samples)]
+          [#:alpha alpha (real-in 0 1) (color-field-alpha)]
+          ) renderer2d?]{
+Returns a renderer that draws rectangles filled with a color based on the center point.
+
+@interaction[#:eval plot-eval
+                    (plot (color-field (λ (x y) (if (< (+ (sqr x) (sqr y)) 1) (random 10) 'black))
+                                        -2 2 -2 2))]
+}
```
Comment (alex-hhh) [inline plot-doc/plot/scribblings/renderer2d.scrbl:214]: Can you please add a history entry here.


```
@history[#:added "7.8"]
```

File: plot-lib/plot/private/plot2d/color-field.rkt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    [(and y-min (not (rational? y-min)))  (fail/pos "#f or rational" 3)]
+    [(and y-max (not (rational? y-max)))  (fail/pos "#f or rational" 4)]
+    [(or (> alpha 1) (not (rational? alpha)))  (fail/kw "real in [0,1]" '#:alpha alpha)]
+    [else
+     (let ([f ((inst fix-a-field-fun Plot-Color) 'color-field f)])
+       (renderer2d (vector (ivl x-min x-max) (ivl y-min y-max)) #f default-ticks-fun
+                   (color-field-render-fun
+                    f samples alpha)))]))
+
+#;(module+ test
```
Comment (alex-hhh) [inline plot-lib/plot/private/plot2d/color-field.rkt:77]: I prefer not to have commented out code in the code base.  Please remove the two tests from this file.  If you want you can add the Mandelbrot set to the unit tests but I also added the sample code with the image to the description of this SR.

**Review by alex-hhh — COMMENTED** [review]
This pull request looks good to me.  I added some comments for adding history entries to the documentation and removing some commented out code, once you address those, I will merge the changes.

**Pull request merged**